docs/man/man1/elbpcom.1
.\" Copyright (C) 2015 Jeff Epler
.TH LinuxCNC "1" "2015-04-18" "LinuxCNC Documentation" ""
.SH NAME
elbpcom \- Communicate with Mesa ethernet cards
.SH SYNOPSIS
Common options:
.RS
.SY elbpcom
.BI [\-\-ip= IP ]
.BI [\-\-port= PORT ]
.BI [\-\-timeout= TIMEOUT ]
.BI [\-\-space= MEMSPACE ]
.BI [\-\-size= TRANSFER_SIZE ]
.RE

Reading data:
.RS
.SY elbpcom
.BI [ common\ options ]
.B [\-\-info]
.BI \-\-address= ADDRESS
.BI \-\-read= LENGTH
.RE

Writing data:
.RS
.SY elbpcom
.BI [ common\ options ]
.BI \-\-address= ADDRESS
.BI \-\-write= HEXDATA
.RE

Read and decode memory space info area:
.RS
.SY elbpcom
.BI [\-\-space= MEMSPACE ]
.BI \-\-read-info
.RE

Sending arbitrary packets:
.RS
.SY elbpcom
.BI [ common\ options ]
.I HEXDATA
.RE

.SH DESCRIPTION
Read or write data from a Mesa ethernet card that uses the LBP16 protocol,
such as the 7I80.  This can be useful for performing certain low-level tasks.

For more information about the meaning of each address space, see the card
documentation.  Incorrect use of this utility can have negative effects such
as changing the board's IP address or even corrupting the FPGA bitfile in
the eeprom.  For some tasks, such as updating FPGA bitfiles and setting IP
addresses,
.BR mesaflash (1)
is a more appropriate tool.

If not specified, the default values are
.RS
.BI \-\-ip= 192.168.1.121
.BI \-\-port= 27181
.BI \-\-timeout= .2
.BI \-\-space= 0
.BI \-\-size= 0
.RE 

If the
.B --size
argument
.I TRANSFER_SIZE
is 0, elbpcom will look up the preferred transfer size of the space in the space's info area.

This example demonstrates reading the HOSTMOT2 identifying string from the
IDROM in space 0:
.RS
.nf
$ elbpcom \-\-address 0x104 \-\-read 8
> 82420401
< 484f53544d4f5432
      HOSTMOT2
.fi
.RE
First the request is shown in hex.  Then the response (if any) is shown in hex.
Finally, the response is shown in ASCII, with "." replacing any non-ASCII
characters.  This is similar to the following invocations of mesaflash:
.RS
.nf
$ ./mesaflash  \-\-device 7i80 \-\-rpo 0x104
54534F48
$ ./mesaflash  \-\-device 7i80 \-\-rpo 0x108
32544F4D
.fi
.RE
but notice its different treatment of byte order.

.SH SEE ALSO
.BR mesaflash (1),
.BR hostmot2 (9),
.BR hm2_eth (9),
.UR http://www.mesanet.com
Mesa's documentation for the Anything I/O boards
.UE .


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/utils/elbpcom.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#    Copyright 2012-2015
#    Peter C. Wallace <pcw@mesanet.com> and Jeff Epler <jepler@unpythonic.net>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import optparse
import re
import readline
import socket
import struct

parser = optparse.OptionParser("%prog [options] [commands]",
    description="Communicate with Mesa ethernet cards using the LBP16 protocol")
parser.add_option("-i", "--ip", dest="sip",
    help="IP address of board (default: 192.168.1.121)",
    metavar="X.Y.Z.W", default="192.168.1.121")
parser.add_option("-p", "--port", dest="sport",
    help="UDP port of board (default: 27181)", type="int", default=27181)
parser.add_option("-t", "--timeout", dest="timeout",
    help="Response timeout in seconds (default: 0.2)", type="float", default=.2)
parser.add_option("-s", "--space", dest="space", default=0,
    choices = ["0", "1", "2", "3", "4", "5", "6", "7"],
    help="Address space to read or write (default: 0, the hm2 register file)")
parser.add_option("--info", dest="info", action="store_true",
    default=False,
    help="Select info area for read operation (default: memory space)")
parser.add_option("-a", "--address", type="int", dest="address", default=None,
    help="Base address to read or write")
parser.add_option("-I", "--increment", dest="increment", action="store_true",
    default=True,
    help="auto-increment address (enabled by default, use --no-increment to disable)")
parser.add_option("-n", "--no-increment", dest="increment",
    action="store_false",
    help="do not auto-increment address")
parser.add_option("--size", dest="size",
    help="Transfer size in number of bytes (default: look up preferred transfer size in the space's info area)",
    type="int", default=0)
parser.add_option("-r", "--read", type="int", dest="read", default=None,
    help="Number of bytes to read")
parser.add_option("-w", "--write", type="string", dest="write", default=None,
    help="Hex-coded Values to read")

parser.add_option("--read-info", dest="read_info", default=False,
    action="store_true",
    help="Read and decode the info area of the selected memory space.")

options, args = parser.parse_args()
if options.space: options.space = int(options.space)

sizemap = {1: 0, 2: 1, 4: 2, 8: 3}

s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
s.settimeout(options.timeout)

def transact(sdata:bytes, quiet=False, response=True):
    s.sendto(sdata, (options.sip, options.sport))
    if not response: return
    try:
        data, daddr = s.recvfrom(1280)
        if not quiet: print("<", data.hex())
        if not quiet: print("     ", re.sub('[^ -~]', '.', data.decode('utf-7', errors='replace')))
        return data
    except socket.timeout:
        if not quiet: print("! no response")
        return None

def interact():
    try:
        while 1:
            sdata = input("> ")
            if not sdata.strip(): break
            transact(sdata)
    except KeyboardInterrupt:
        pass

def make_read_request(space:int, info:bool, size:int, increment:bool, address:int, nbytes:int) -> bytes:
    return struct.pack("<HH",
        (1<<14) | (space << 10) | (info << 13)
        | (sizemap[size] << 8) | (increment << 7) | int(nbytes/size),
        address)

def make_write_request(space:int, info:bool, size:int, increment:bool, address:int, data:bytes) -> bytes:
    return struct.pack("<HH",
        (1<<15) | (1<<14) | (space << 10) | (info << 13)
        | (sizemap[size] << 8) | (increment << 7) | int(len(data) / size),
        address) + data

def optimal_size(space:int, info:bool, address:int, nbytes:int):
    if info: return 2
    info = transact(make_read_request(space, True, 2, True, 2, 4), quiet=True)
    if info is None:
        raise RuntimeError("Failed to get information about memory space %d" % space)
    memsizes, memranges = struct.unpack("<HH", info)
    maxaddr = 1 << (memranges & 0x3f)
    #print "# Note: space %d has maxaddr %d (memsizes=0x%x memranges=0x%x)" % (space, maxaddr, memsizes, memranges)
    # This gets the wrong limit with my firmware on space 6, so disabling it
    #if address + nbytes >= maxaddr: raise ValueError, "Address out of range (address=%d nbytes=%d maxaddr=%d)" % (address, nbytes, maxaddr)
    for i in (3,2,1,0):
        b = (1<<i)
        if address % b or nbytes % b or not memsizes & b: continue
        return b
    raise ValueError("Access size incompatible with address or length (address=%d nbytes=%d memsizes=%d)" % (address, nbytes, memsizes))

def get_uint16(data:bytes, offset:int):
    return (data[offset+1] << 8) | data[offset]

def print_info(data:bytes):
    print(f"info area for memory space {options.space}:")

    cookie = get_uint16(data, 0)
    print(f"    cookie: 0x{cookie:04x}")

    memsize = get_uint16(data, 2)
    memsize_writable = memsize & 0x8000
    memsize_type = (memsize & 0x7f00) >> 8
    memsize_access = memsize & 0x000f
    print(f"    memsize: 0x{memsize:04x}")

    print("        Writable" if memsize_writable else "        Read-Only")

    if (memsize_type == 0x01):
        print("        type=01 (Register)")
    elif (memsize_type == 0x02):
        print("        type=02 (Memory)")
    elif (memsize_type == 0x0e):
        print("        type=0e (EEPROM)")
    elif (memsize_type == 0x0f):
        print("        type=0f (Flash)")
    else:
        print(f"        type={memsize_type:02x}(unknown)")

    print(f"        access=0x{memsize_access:01x}", end='')
    if memsize_access & 0x1:
        print(" 8-bit", end='')
    if memsize_access & 0x2:
        print(" 16-bit", end='')
    if memsize_access & 0x4:
        print(" 32-bit", end='')
    if memsize_access & 0x8:
        print(" 64-bit", end='')
    print()

    memranges = get_uint16(data, 4)
    e = (memranges & 0xf800) >> 11
    p = (memranges & 0x07c0) >> 6
    s = memranges & 0x003f
    print(f"    memranges: 0x{memranges:04x}")
    print(f"        erase block size: {e} ({2**e} bytes)")
    print(f"        page size: {p} ({2**p} bytes)")
    print(f"        Ps address range: {s} ({2**s} bytes)")

    addr_ptr = get_uint16(data, 6)
    print(f"    addr ptr: 0x{addr_ptr:04x}")

    name = str(data[8:], encoding='utf-7')
    print(f"    name: {name}")


if options.read:
    if options.address is None: raise SystemExit("--read must specify --address")
    if options.size == 0:
        options.size = optimal_size(options.space, options.info, options.address, options.read if options.increment else 0)
    command = make_read_request(options.space, options.info, options.size, options.increment, options.address, options.read)
    print(">", command.hex())
    transact(command)

elif options.write:
    if options.address is None: raise SystemExit("--write must specify --address")
    write = bytes.fromhex(options.write)
    if options.size == 0:
        options.size = optimal_size(options.space, options.info, options.address, len(write) if options.increment else 0)
    command = make_write_request(options.space, options.info, options.size, options.increment, options.address, write)
    print(">", command.hex())
    transact(command, response=False)

elif options.read_info:
    if options.size == 0:
        options.size = optimal_size(options.space, True, 0x0000, 16)
    command = make_read_request(options.space, True, options.size, options.increment, 0x00, 0x10)
    print(">", command.hex())
    data = transact(command)
    print_info(data)

elif args:
    for a in args:
        transact(a)
else:
    interact()
